### ISSUE TITLE
batch solution of N small linear systems under the same MPI communicator?

### ISSUE THREAD

**Issue by p-costa** [hypre-space/hypre#332]
Dear all,

I am solving a Poisson equation of this kind, for variable $\Phi$ with $f$ the right-hand-side vector using the `Struct` interface, `PFMG`:

![eq1](https://user-images.githubusercontent.com/29615530/114690719-82843880-9d06-11eb-863b-3aeb671ed600.png)
 
However, to speedup the solution, I use a FFT-based expansion to decouple the equations in one homogeneous direction (here the direction pertaining to index `j`), resulting in the following Helmholtz equation:

![eq2](https://user-images.githubusercontent.com/29615530/114690768-8ca63700-9d06-11eb-8e08-e371854a688a.png)

Where I need to have the tasks distributed in a pencil-like decomposition so that the FFTs in one direction can be performed (figure below, left). Then I solve $N$ independent 2D linear systems along the homogeneous direction where I perform this FFT-based reduction (using the `Struct` interface, `PFMG`).
 
I get the expected major speedup for a small number of cores. However, each of these small 2D systems has to be solved over `MPI_COMM_WORLD`, and tasks have to sync between each solve, which compromise the parallel scalability. My solution right now, which works like a charm, is to redistribute the tasks in a slab-like domain decomposition as illustrated in the following figure, and solve these $N$ 2D systems with the parallelization of the right, over a new MPI communicator encapsulating the tasks responsible for each of the small systems. This way the solution of the independent 2D systems is distributed.

![all-to-all](https://user-images.githubusercontent.com/29615530/114690808-962f9f00-9d06-11eb-81ee-b279c8d0dd52.png)
 
This redistribution, however, involves an *all-to-all* collective operation (actually two, since I need to perform the backward FFT-based transform to get back the solution), which is something that I feel could be avoided if a batch solution of $N$ systems would be supported by *HYPRE*. (Also, I have a hard limit of the number of tasks I can decompose among *slabs* dictated by the number of grid cells in that direction.)
 
So after this long explanation, which I hope was clear, my short question is -- do you see it feasible/likely that *HYPRE* will support a batch solution of this kind in the future for use cases like this? Or perhaps there is a workaround that can be explored using the current API?
 
Thank you in advance!
Pedro

**Comment by rfalgout** [comment]
Hi @p-costa .  I want to make sure I understand what you have here.  It sounds like you've got N distinct 2D problems in the x1-x3 plane of your left figure (so, they are not coupled in the x2 direction) that you want to solve.  If you solve them one at a time, then you pay a communication latency cost that grows with N, so you would prefer to solve them "simultaneously" and aggregate the communication so as to reduce the latency overhead.

If that is correct, then one thing you could try now is to just setup up a big 3D system with the original decomposition in the left figure and set the stencil coupling(s) in the x2 direction to zero.  This would incur some communication in x2, but PFMG may still work well here (do you have any anisotropy in the 2D problems?) and I think this would scale better.  There might be a way to explicitly eliminate the stencil entry in the x2 direction, but I'm not confident the current code would work correctly right now.  I could give it a try pretty easily, but let me know first if I understand things correctly.  Thanks!

**Comment by p-costa** [comment]
Hi @rfalgout,

Thank you for your reply!

Yes, you have understood my problem correctly. I am in fact solving a 3D Poisson equation on a grid that is structured, but not necessarily uniform in the directions where I solve each 2D system (i.e., the equation above is simplified for constant grid spacing in each direction). It just needs to be uniform along the direction where this decoupling using fast Fourier/cosine/sine transforms is performed.

I remember that I tried removing the stencil couplings in that direction (zeroing two of 7-point stencil coefficients), and that I also tried to set up a 5 point-stencil for the full 3D system. I think that both options resulted in a fatal error, although I do not remember precisely the problem I got (either SIGFPE or SEGFAULT). I can confirm this tomorrow.

**EDIT:** To be clear, the coefficient matrix will vary along direction x_2, but only the diagonal elements (according to that parameter \lambda_j shown in the second equation).

Thanks again for the suggestion!
Pedro

**Comment by p-costa** [comment]
Hi again, @rfalgout and all,

Actually, I reverted the code to the implementation I had, where I set the stencil couplings in the 'FFT direction' to zero, and now it actually works. I get the correct solution for a test performed on my workstation!

However, the simulation wall-clock per step for solving the 3D system coupled in all directions is roughly the same as that of the 3D system with zeroed stencil couplings. The measured wall-clock time per step of the code on a small system, 4 cores, (it is a new multi-block finite-difference Navier-Stokes solver), is as follows:

1. standard 3D system with *no* FFT pre-processing -- **3.1 seconds**
2. FFT pre-processing and 3D system with zeroed stencil couplings -- **3.5 seconds**
3. FFT pre-processing and N, 2D systems -- **0.13 seconds** ~ 25 times faster!

Hence, even though I obtain the correct solution, I seem to be very far from the performance gain I could achieve by explicitly decoupling the systems. Is this expected? I submitted a more relevant large-scale test on a supercomputer and will update this issue with a strong scaling plot for the different approaches.

I should note that, for larger systems on a many-core calculation, I typically obtain a speedup of 2-10, depending on the problem, not 25. Also, I edited the comment above with a note, which I write again here for clarity: the coefficient matrix will vary along direction x_2, but only the diagonal elements (according to that parameter \lambda_j shown in the second equation).

Thank you for your feedback,
Pedro

---

**EDIT:** I'm sorry, my measurements for this small case were incorrect (I was accidentally comparing SMG with PFMG, and the right-hand-side was also poorly chosen), for future reference I write here the correct values:

1. standard 3D system with *no* FFT pre-processing -- **0.79 seconds**
2. FFT pre-processing and 3D system with zeroed stencil couplings -- **0.75 seconds**
3. FFT pre-processing and N, 2D systems -- **0.19 seconds**

This factor of 4 is also more consistent with what I typically observe.

**Comment by rfalgout** [comment]
Hi Pedro,

There could be a few reasons why you don't see a performance gain, but I need more details, especially about convergence and problem size.

How many iterations does each of the above approaches (1, 2, and 3) take per solve?  In case 3 with the N 2D systems, how many iterations does each of the 2D solves take?  You mention the diagonal changing in each of these solves, which makes me wonder if some or many of the systems are easy to solve.  In the 3D case (approach 2), the convergence should be the same as (or similar to) the slowest 2D solve, so it could lead to lots of unnecessary work.

The problem size and parallel distribution information are other factors that could make a big difference.  For the problem you tested, was there any parallelism in the individual 2D solves for approach 3, or were they each entirely sequential.  Communication is the dominant cost in PFMG solves, especially for smaller problems.  If you are getting perfect parallelism for the 2D solve part of approach 3, then the all-to-all communication can be worth it (especially for smaller problems).

Thanks!

**Comment by p-costa** [comment]
Hi Rob,

Thank you very much! Your explanation for what is possibly happening makes perfect sense to me. I will have HYPRE logging that information and follow up here, also with the parallel performance of the 3D system with decoupled stencil elements.

Here is a figure showing the scaling of approaches 1 and 3, where approach 3 is tested with and without that _all-to-all_ collective, for two different cubic boxes with uniform grid (when that is the case I also use `HYPRE_StructMatrixSetSymmetric`). The number of grid points is 512^3 and 1024^3.

<img width="440" alt="Screenshot 2021-04-15 at 15 06 29" src="https://user-images.githubusercontent.com/29615530/114892307-2b608f80-9dfc-11eb-805f-820885160c34.png">

The scaling is excellent for the standard 3D solve (green) and the N, 2D solves with all-to-all (blue), while it deteriorates badly with N, 2D solves without all-to-all (orange). Different shades of the same color correspond to the pencils being aligned in direction `x_1` (darkest), `x_2`, or `x_3` (lightest). The blue lines cannot go more to the right because I run out grid points to decompose among slabs, while the green line for the largest domain and lowest number of processes yielded an out-of-memory error.

Thanks again for your feedback, and I will follow up when I have information about the number of iterations of PFMG, and the scaling of approach 2 above.

Pedro

**Comment by p-costa** [comment]
Hi again Rob,

I now have more detailed information. I tested the incompressible Navier-Stokes solver on a single box with 512x512x512 grid points.

The number of iterations of PFMG for the same prescribed tolerance is:

| 3D PFMG | 3D PFMG (zeroed stencil couplings) | N, 2D PFMG                 |
|:-------:|:----------------------------------:|----------------------------|
|     ~ 15|                               ~ 13 |  **max:** 13, **min:** 6   |

And the following figure shows a strong scaling plot (cumulative wall-clock time for three solutions):

<img width="531" alt="strong_scaling" src="https://user-images.githubusercontent.com/29615530/115128598-716d4d00-9fce-11eb-8b0a-b33968f21e3d.png">

Summarizing the observations:

1. FFT-based pre-processing without data redistribution (*all-to-all*) will, expectedly, not scale;
2. FFT-based pre-processing with *all-to-all* gives a consistent 2x speedup by allowing me to solve batches of 2D slabs in parallel. However, I cannot go beyond 512 CPUs here, because I run out of grid points to decompose among slabs;
3. FFT-based pre-processing with 3D PFMG, decoupled in one direction, yields roughly the same wall-clock time per step as the standard solve.

Based on your feedback, what I observe in point 3 seems to make sense. The magnitude of the diagonal coefficients of the 2D systems increases monotonically between each solve, due to the parameter λ above (varies from 0 to -4 for a Dirichlet--Dirichlet boundary condition in this case). Accordingly, the PFMG iterations decrease monotonically: the higher the diagonal coefficients, the fastest PFMG converges.

I should note that here this FFT-based synthesis of the equation yields a speedup of 2x, but I observed up to 8x speedup in different configurations (the larger the number grid points along the 'FFT direction', the larger the speedup).

So perhaps the only option to avoid the *all-to-all* collective to improve scalability (especially when the solver is ported to GPUs) would be to solve a batch of independent 2D systems, asynchronously. Do you agree with my analysis?

Thanks again for your feedback!

Pedro

---

Some minor points, just for reference:

- The figure shows the averaged wall-clock time per step for the overall solution, where the Poisson equation is solved three times, corresponding to ~99% of the simulation time. The code ran on the Swedish supercomputer *Tetralith*.

- The multi-block Navier-Stokes solver can be found [here](https://github.com/p-costa/SNaC/tree/multi_block_fft_mpi_f08_transpose), and the subroutines using *HYPRE* [here](https://github.com/p-costa/SNaC/blob/multi_block_fft_mpi_f08_transpose/src/solver.f90).

- The 25-fold speedup I reported from a small case on my WS, in a comment above, was due to accidentally comparing PFMG with SMG; I edited the comment accordingly.

- Reference describing the method I'm using to decouple the systems of equations: Schumann, and Sweet. *Fast Fourier transforms for direct solution of Poisson's equation with staggered boundary conditions*.  J. Comput. Phys. 75.1 (1988): 123-137.

**Comment by rfalgout** [comment]
Hi Pedro,

Given the different iteration counts for the 2D systems, one thing that could help quite a bit is to break your big 3D-FFT-transformed system up into multiple smaller 3D systems that have similar 2D problems within them.  For example, you might form a hard (10 -13 iterations per 2D subsystem) and an easy (6 - 9 iterations per 2D subsystem) 3D system to solve one after the other with PFMG (with zeroed coupling in one direction).  You could break it up even more if it helps.

One question I didn't see an answer to was whether your 2D systems are fully sequential in the all-to-all N x 2D PFMG case.  I am guessing this is sometimes true, but not always?

Thanks!

-Rob

**Comment by p-costa** [comment]
Hi Rob,

Thanks for the suggestion! Given the distribution of number iterations along direction `x_2` (stair-step-like with several planes taking roughly the same number of iterations), this seems like a promising solution! When I find time to test it, I will report back here.

> One question I didn't see an answer to was whether your 2D systems are fully sequential in the all-to-all N x 2D PFMG case. I am guessing this is sometimes true, but not always?

I realize that I missed that question, my apologies. In the all-to-all N x 2D PFMG case, batches of 2D systems are solved in parallel. For the single 512^3 box example, each plane is solved by only one task, while for my original sketch each 2D plane is solved by four tasks. I will explain in more detail what I'm doing:

In the sketch of my original post (I should have explained it better), the different colors denote different blocks, with which I described my geometry, while different shades of the same color denote the individual MPI tasks. The *all-to-all* collective is performed for the tasks within each block (i.e. within a communicator `comm_block`, not `MPI_COMM_WORLD`), and as you can see the boundaries between different blocks are not congruent anymore (for instance, the green slabs are much more decomposed than the orange).

Therefore, to solve the 2D systems in parallel, I need to define an array of communicators which contain the tasks responsible for each plane. Perhaps it becomes clearer with the following snippet from my Fortran implementation:
```.f90
do i=lo_idir,hi_idir ! loop within local slab-normal points, direction x_2 above
  n = i-lo_s_idir+1
  icolor = MPI_UNDEFINED
  if( i >= lo_s_idir .and. i <= hi_s_idir) icolor=i
  ! extract subset of MPI_COMM_WORLD responsible for 2D plane i:
  call MPI_COMM_SPLIT(MPI_COMM_WORLD,icolor,myid,comm)
  ! MPI communicator to be passed to HYPRE:
  if(i >= lo_s_idir .and. i <= hi_s_idir) comms_slab(n) = comm
enddo
```
Then, for each of the slabs in the sketch of the right, I solve sequentially 2D planes (`j = 1` to `j = n_2` with `n_2` the number of grid points along `x_2`, within the slab), by passing to *HYPRE* `comms_slab(j)` when setting up each system.

So, in short, it is parallel within `MPI_COMM_WORLD`, and in the original sketch each 2D system is solved by *HYPRE* using 4 different tasks (different colors of the figure), and it is sequential within each of the slabs. I hope my explanation is clear.

Thanks again!
Pedro

**Comment by p-costa** [comment]
Hi again, Rob and all,

I still did not have the time to test the solution you [suggested](https://github.com/hypre-space/hypre/issues/332#issuecomment-822827121), but I now know the issue I had a long time ago when I tried zeroing the stencil couplings along one direction. Quoting my previous comment:

> I remember that I tried removing the stencil couplings in that direction (zeroing two of 7-point stencil coefficients), and that I also tried to set up a 5 point-stencil for the full 3D system. I think that both options resulted in a fatal error, although I do not remember precisely the problem I got (either SIGFPE or SEGFAULT).

What happens is that when I compile my program with flags to capture floating-point exceptions and abort the program, my program aborts at the line corresponding to the setup of the solver: `call HYPRE_StructPFMGSetup(solver,mat,rhs,sol,ierr)`.

More specifically, the program aborts when compiled with gfortran's MPI wrapper with these flags `-O0 -g -fbacktrace -finit-real=snan -ffpe-trap=invalid`, and linking the library built with GNU compilers:

```
Program received signal SIGFPE: Floating-point exception - erroneous arithmetic operation.
```

the same happens if I compile with the intel MPI wrapper (with the equivalent flags `-O0 -g -traceback -fpe0`, and linking the library built with intel compilers):

```
forrtl: error (65): floating invalid
```
However, if I do not compile my program with these flags, the simulation runs fine and I get the right results. Is this expected?

Also, to be clear, if I do not set the stencil couplings to zero, the program compiled with those debugging flags to trap SIGFPEs will run fine.

**Comment by rfalgout** [comment]
Hi Pedro,

That's good to know.  I will try to look into this at some point and fix.  Do you happen to know where (line number and file) the floating-point exception occurs in the code?  Thanks!

-Rob

**Comment by p-costa** [comment]
Hi Rob,

Sure. I had to configure my build as follows: `export CFLAGS='-traceback' CXXFLAGS='-traceback' FCFLAGS='-traceback' && ./configure --enable-debug` (using Intel oneAPI compilers).

And the exception occurs in line 180 of `pfmg_setup.c`:

```
forrtl: error (65): floating invalid
Image              PC                Routine            Line        Source             
snac               00000000005CB0BB  Unknown               Unknown  Unknown
libpthread-2.33.s  00007FDC7ED05960  Unknown               Unknown  Unknown
snac               00000000004AB1C1  hypre_PFMGSetup           180  pfmg_setup.c
snac               00000000004A58DB  hypre_structpfmgs          58  F90_HYPRE_struct_pfmg.c
```
(corresponding to branch master, commit b3a4a76a5fc965b148b55ed8e9c342666b9c3c41).

edit, best configure with: `export CFLAGS='-g -traceback' CXXFLAGS='-g -traceback' FCFLAGS='-g -traceback' && ./configure`

**Comment by rfalgout** [comment]
Thanks!  This should be easy to fix.

**Comment by p-costa** [comment]
Thank you! I can also now follow up on your suggestion:

> one thing that could help quite a bit is to break your big 3D-FFT-transformed system up into multiple smaller 3D systems that have similar 2D problems within them.

I have tried this, and indeed it helps. I sliced evenly the big 3D system into smaller parts (for a box with 512^3 points I tested 8, 16, 32, and 64 partitions), and I managed to get closer to the performance I got when using the all-to-all collective. The "N x 2D with all-to-all" approach is still about 10%--20% faster than the best-performing partitioning (here 16), but perhaps if I partition the systems unevenly, based on the distribution of maximum number of iterations, I can get closer or even better.

Thanks again!
Pedro

**Comment by rfalgout** [comment]
Hi Pedro,

Would you mind testing the code in branch `pfmg-fix` to see if it fixes your divide-by-zero problem?  The fix was simple, but I was unable to replicate the issue over here, since floating-point divide-by-zero is technically legal in IEEE.  I probably could have found a way eventually, but I gave up. :)  If you can, I would appreciate it.  Thanks!

-Rob

**Comment by p-costa** [comment]
Hi Rob,

Thank you very much for fixing this so quickly! That fixes that divide-by-zero problem (or maybe 0/0?).

However, I now get a `floating overflow` error in line 235:

```C
            alpha += 1.0/(dxyz[d]*dxyz[d]);
```
and if I naively change the code to avoid the exception (`alpha += 1.0`), I get a second `floating overflow` in line 254:

```C
                     beta += 1.0/(dxyz[d]*dxyz[d]);
```
and if I, again, naively change the code to `beta += 1.0`, the code runs and I get the correct results.

Thanks again!
Pedro

**Comment by rfalgout** [comment]
Thanks, Pedro!  What is the size of the problem you are running?  I will try to see if I can replicate/fix on our end.

**Comment by p-costa** [comment]
Hi Rob,

I replicated now the issue on a small box with 64x64x64 grid points. I added the following printing statement above the line:

```C
printf("dxyz[0] = %e dxyz[1] = %e dxyz[2] = %e, d = %d \n",dxyz[0],dxyz[1],dxyz[2],d);
```
and I get the following output:
```
dxyz[0] = 1.000000e+00 dxyz[1] = 1.000000e+00 dxyz[2] = 1.797693e+305, d = 0 
dxyz[0] = 1.000000e+00 dxyz[1] = 1.000000e+00 dxyz[2] = 1.797693e+305, d = 1 
dxyz[0] = 1.000000e+00 dxyz[1] = 1.000000e+00 dxyz[2] = 1.797693e+305, d = 2 
forrtl: error (72): floating overflow
Image              PC                Routine            Line        Source             
snac               00000000005DBBFB  Unknown               Unknown  Unknown
libpthread-2.33.s  0000147B58771960  Unknown               Unknown  Unknown
snac               00000000004A2E1B  hypre_PFMGSetup           236  pfmg_setup.c
snac               00000000004A0041  HYPRE_StructPFMGS          39  HYPRE_struct_pfmg.c
snac               000000000049D841  hypre_structpfmgs          58  F90_HYPRE_struct_pfmg.c
```
In this example I set the stencil couplings to zero along the third direction. If you need the details about the boundary conditions, please let me know.

Hope this helps! I can find time to make a small reproducer in Fortran if needed.

Thank you!
Pedro

**Comment by rfalgout** [comment]
Thanks!  This should be enough to figure out what's going on.  Stay tuned.

**Comment by p-costa** [comment]
Hi Rob,

Now it works for me. Thank you!

Pedro

**Comment by rfalgout** [comment]
Great!  Thanks for helping me out on this.  This was not a case we ever tested, but even if we had, the compiler in our usual development environment doesn't complain anyway. :(

Thanks again!  This will be pushed to the master soon and will be in the next release as well.

**Comment by p-costa** [comment]
Thank you, Rob!

I am happy that I helped a bit, although nothing compared to how much *HYPRE* has been useful to me and my colleagues. In addition to the performance, I personally love the API.

As far as I know, these flags to capture floating-point exceptions, often combined with flags to initialize variables as `NaN` to detect usage of uninitialized variables, are not uncommon in Fortran projects built in 'debug mode'.
